### ISSUE TITLE
setClientId() not specified receives a null value

### ISSUE THREAD

**Issue by yanict** [fourlabsldn/GampBundle#1]
Hi,

I noticed I had a lot of log entries with the following message :
` Uncaught PHP Exception InvalidArgumentException: "For Analytics object, you must specify a value to be set for setClientId"`

Also, after looking for a while in my code, I found that there was an error in the **AnalyticsFactory** class on line 25.
`->setClientId($request->cookies->get('_ga', null))`

Based on the **_Google Analytics Measurement Protocol library for PHP_** the **setClientId **method need a value and cannot receive a **null**.

Am I right and does anybody also had this problem ?
Thanks for your advice.

**Comment by derpue** [comment]
To my information the client ID is also not the whole _ga cookie, instead its somthing like `explode(".", $_COOKIE['_ga'])[2]`

**Comment by yanict** [comment]
indeed you where right... almost ;)
I found this : https://plus.google.com/u/1/110147996971766876369/posts/Mz1ksPoBGHx

basically we need to pass the value after the second dot.
for example, if the _ga value is 
`1.2.299259584.1357814039257`
you will have to do 
`->setClientId(299259584.1357814039257)`

PHP code is in the link I pasted if you need.

**Comment by foaly-nr1** [comment]
Any chance one of you can open a PR? @derpue @yanict

**Comment by derpue** [comment]
@yanict Yes you are right, both the random ID (3rd part) and the timestamp the cookie was created (4th part) are together the clientId. Just did a test comparing the cookie and the value returned from the GA Tracker object in Javascript.

@all How should the exact behavior be defined?
I mean, this library is normally used to track server side events (eg a REST Interface) and not a normal user browsing the page (as this is all done by the normal GA client-side Javascript), thus we normally don't have a _ga Cookie, most of the time also we will set our own cookie, if we don't have a way to identify the user by eg login credentials.

Should this bundle handle all automagically? Test for _ga cookie, if not set, create own clientId, set own cookie and so on (needs also all be configurable).
Or do it as it's now and use the _ga cookies clientId if present just for convenience, but do a more explicit documentation of that clientId issue?

As the clientId is mandatory, but can't be set in parameters as the other mandatory values, make it more explicitly clear by changing the factory to somewhat `$analytics = $container->get('gamp.factory')->create($clientId)`

Dunno whats the best way

**Comment by foaly-nr1** [comment]
@derpue Leaving it at @Nightbr's implementation for now.

**Comment by Nightbr** [comment]
I'm not reopening this issue but here we are with this:

- My solution which retrieves automatically the _ga cookie works fine with a classic website which set the cookie...

- But if you have a REST API (also my case), there is no _ga cookie and it throws an error beacause I haven't the clientId parameters...

For now, I workaround this by setting the clientId equals to the userId `$this->analytics->setClientId($user->getId());`

For example, the service Segment do the same thing:

>  If you do not pass this through, we look for the userId or anonymousId value and set it as the cid.

<https://segment.com/docs/integrations/google-analytics/>

But it is not ideal... you need to send event only if user is in context (you can send event with anonymous user without the _ga cookie)...

One solution can be to generate the cookie and set it into the header when the clientId is null. But it will work for application which handles cookies and read the header set-cookie... And if the client application of our REST service doesn't manage cookie, this solution will generate one clientId to each request and it will generate a new user each time... 

I continue thinking about this... but one solution can be to stop reporting Analytics through the backend and use only JavaScript to report data... Develop webapp and no more website...
